### DISCUSSION TITLE
EIP-1898 Support

### DISCUSSION THREAD

**Discussion by infloop** [wevm/viem#3165] [Idea / Feature Request] [answered]
Hi! 

Do you have plans to have support for EIP-1898 (https://eips.ethereum.org/EIPS/eip-1898)? I'd like to create a PR for that.

Although this EIP is marked as stagnant, it is widely supported by most clients and actively used in production to retrieve deterministic state from Ethereum RPCs. It enables clients to unambiguously specify the block they want to query for certain RPC methods.

RPC methods affected by EIP-1898:

```
eth_getBalance
eth_getStorageAt
eth_getTransactionCount
eth_getCode
eth_call
eth_getProof
```

### Examples

Before EIP-1898:

```json
{
  "method": "eth_getBalance",
  "params": ["0x8D97689C9818892B700e27F316cc3E41e17fBeb9", "0xf65631529d476553ca5b0056d6480c3970dd5ac884fee51d5b30ca7fceab8894"],
  "id": "42",
  "jsonrpc": "2.0"
}
```

```json
{
  "method": "eth_getBalance",
  "params": ["0x8D97689C9818892B700e27F316cc3E41e17fBeb9", "0x1482fbb"],
  "id": "42",
  "jsonrpc": "2.0"
}
```

After EIP-1898:
```json
{
  "method": "eth_getBalance",
  "params": ["0x8D97689C9818892B700e27F316cc3E41e17fBeb9", {
    "blockHash": "0xf65631529d476553ca5b0056d6480c3970dd5ac884fee51d5b30ca7fceab8894"
  }],
  "id": "42",
  "jsonrpc": "2.0"
}
```

```json
{
  "method": "eth_getBalance",
  "params": ["0x8D97689C9818892B700e27F316cc3E41e17fBeb9", {
    "blockNumber": "0x1482fbb"
  }],
  "id": "42",
  "jsonrpc": "2.0"
}
```

### Usage

```typescript
const url = '...';
const transport = http(url);
const client = createPublicClient({
  chain: mainnet,
  transport: transport,
});

const balance = await client.getBalance({ 
  address: '0x8D97689C9818892B700e27F316cc3E41e17fBeb9', 
  blockTag: { blockHash: '0xf65631529d476553ca5b0056d6480c3970dd5ac884fee51d5b30ca7fceab8894' }
}),
```
### Extra

Clients and services that support EIP-1898:
- Geth: https://github.com/ethereum/go-ethereum/pull/19491
- Besu: https://github.com/hyperledger/besu-docs/issues/1242
- Erigon: https://github.com/erigontech/erigon/pull/2736
- Nethermind (except blockNumber): https://github.com/NethermindEth/nethermind/issues/3426, https://github.com/NethermindEth/nethermind/issues/5025
- Infura: https://docs.metamask.io/services/reference/scroll/json-rpc-methods/#block-identifier 
- Alchemy: https://docs.alchemy.com/reference/eth-call#parameters

Other links: 
https://github.com/ethers-io/ethers.js/issues/2808

**Comment by sambacha** [comment]
Is this not supported?

**Comment by jxom** [comment, marked as answer]
PR welcome.

**Reply by infloop** [reply]
@jxom PR is done, can you take a look? https://github.com/wevm/viem/pull/4186
